source/common/router/header_parser.cc

```diff
@@ -342,28 +342,53 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
     headers.remove(header);
   }
 
+  // Temporary storage to hold evaluated values of headers to add and replace. This is required
+  // to execute all formatters using the original received headers.
+  // Only after all the formatters produced the new values of the headers, the headers are set.
+  // absl::InlinedVector is optimized for 4 headers. After that it behaves as normal std::vector.
+  // It is assumed that most of the use cases will add or modify fairly small number of headers
+  // (<=4). If this assumption changes, the number of inlined capacity should be increased.
+  // header_formatter_speed_test.cc provides micro-benchmark for evaluating speed of adding and
+  // replacing headers and should be used when modifying the code below to access the performance
+  // impact of code changes.
+  absl::InlinedVector<std::pair<const Http::LowerCaseString&, const std::string>, 4> headers_to_add,
+      headers_to_overwrite;
+  std::string value_buffer;
```

Comment:
nit: not needed.

Reply:
It is needed to avoid superfluous string-copies when stream_info==nullptr

actually I suspect we could also make entry.formatter_->format avoid the string-copy when it's not doing anything, which would make it matter more to have this buffer.

@adisuissa see comment-thread at line 369.

Reply:
Not sure I fully grok, if `stream_info == nullptr` then `value_buffer` won't be used, or am I missing something here? (FWIW, I see the benefit of `value`, but not of `value_buffer`)
In any case, if this is some intentional code pattern, I suggest adding some comments that describe this.

Reply:
This is to have a place to store the computed value from entry.formatter_->format() when stream_info!=null, but avoiding the string alloc+copy when stream_info==null.

I agree that could be added as a comment.